nixos/modules/services/cluster/k3s/default.nix

```diff
@@ -33,15 +33,40 @@ in
 
     serverAddr = mkOption {
       type = types.str;
-      description = lib.mdDoc "The k3s server to connect to. This option only makes sense for an agent.";
+      description = lib.mdDoc ''
+        The k3s server to connect to.
+
+        Servers and agents need to communicate each other. Read
+        [the networking docs](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking)
+        to know how to configure the firewall.
+      '';
       example = "https://10.0.0.10:6443";
       default = "";
     };
 
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
```

Comment:
```suggestion
      default = cfg.role == "server" && cfg.serverAddr == "";
```

Reply:
It seems we have 3 roles: `master`, `replica`, `node`.
"master" won't have `serverAddr` but will have `clusterInit` true.
`replica` is a server but is not master, requires serverAddr and denies use of clusterInit.
Maybe we should change roles to reflect this automatically. And infer from there. Also warn when some field is missing.
`nodes` seek to use all IPs from master & replica in serverAddr.
Do you agree?

Reply:
@euank Do you think `roles` option should account for HA needs?

Reply:
I'm not totally sure.

The "roles" thing is there because the upstream docs and installation instructions talk a lot about "agent" vs "server".

Ideally, I'd like to keep our concepts as close to the upstream k3s docs so that those docs can help someone figure out how to use the nixos module too.

For consistency with [upstream docs](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/), I'd prefer to expose `clusterInit` over adding a new replica role to mask over it.

Reply:
I also think it's better to stick closer to upstream.